internal/kargo/promotion_builder.go

```diff
@@ -0,0 +1,264 @@
+package kargo
```

Comment:
This wasn't introduced by this PR, but this `internal/kargo` package has been bugging me for a while because I don't feel like the package has a clearly defined purpose. (Maybe it does, and I'm failing to see it.)

If I'm not wrong, there's certainly no need to untangle this knot to get this PR merged, but (again, only if I'm right), I wonder if there's any other place that might be appropriate for the new functionality added to the package, in particular `NewPromotionBuilder()`.

Reply:
The only alternative I can think of is inventing a new `promotion` package, but I do not really like this alternative either...

Reply:
We can put a pin in it. Fortunately there are only a couple packages like this that seem to have collected unrelated bits of functionality over time for lack of a better home.